barretenberg/cpp/src/barretenberg/proof_system/circuit_builder/ultra_circuit_builder.cpp

```diff
@@ -451,6 +458,68 @@ template <typename FF> void UltraCircuitBuilder_<FF>::create_ecc_add_gate(const
     q_sort.emplace_back(0);
     q_lookup_type.emplace_back(0);
     q_elliptic.emplace_back(0);
+    q_elliptic_double.emplace_back(0);
+    q_aux.emplace_back(0);
+    ++this->num_gates;
+}
+
+/**
+ * @brief Create an elliptic curve doubling gate
+ *
+ *
+ * @param in Elliptic curve point doubling gate parameters
+ */
+template <typename FF> void UltraCircuitBuilder_<FF>::create_ecc_dbl_gate(const ecc_dbl_gate_<FF>& in)
+{
+    /**
+     * gate structure:
+     * | 1  | 2  | 3  | 4  |
+     * | -  | x1 | y1 | -  |
+     * | -  | x3 | y3 | -  |
+     * we can chain an ecc_add_gate + an ecc_dbl_gate if x3 y3 of previous add_gate equals x1 y1 of current gate
+     * can also chain double gates together
+     **/
+    bool can_fuse_into_previous_gate = true;
+    can_fuse_into_previous_gate = can_fuse_into_previous_gate && (w_r[this->num_gates - 1] == in.x1);
+    can_fuse_into_previous_gate = can_fuse_into_previous_gate && (w_o[this->num_gates - 1] == in.y1);
```

Comment:
Okay so reading this to understand why its done:


A doubling operation will consume two rows, one for the input point and another for the output point. If the next operation is also a double, you can save a row, by setting the previous q_elliptic_double selector to be one and then you only need to add a row for the output point (x3, y3)

Reply:
that is correct. The specific structure of the double gate also allows us to chain an ecc_add_gate and an ecc_double_gate together (i.e. output points of ecc_add_gate == input points of ecc_double_gate)